### DISCUSSION TITLE
best practise apply pint function on pandas dataframe

### DISCUSSION THREAD

**Discussion by SchabiDesigns** [hgrecco/pint#2031] [Q&A] [answered]
Hi guys,

I am quite sure I am not the first who noticed the bad performance when applying a pint function on a large dataframe. :)
So when I have an column with values and want to convert them to another unit with a given function that takes forever.
I am aware of the workaround with **pint-pandas** to add the needed values first as separate columns to the dataframe  and then calculate the desired result column, but I would like to prevent to code another method to adress data input from dataframes.  How do you address that issue?

![image](https://github.com/hgrecco/pint/assets/97088571/e6502799-3e90-4435-ac25-bf8a4670d719)

the function itself is not a big deal...
```
@ureg.wraps("gravity", ("volt", None, "gravity", "volt", "volt"))
def v_2_g(x: ureg.Quantity, gain: float, sensor_range: ureg.Quantity, volt_range: ureg.Quantity = Q_("10 V"), sensor_volt_range: ureg.Quantity = Q_("5 V")) -> ureg.Quantity:
...
```

Thanks for your input in advance! 👍 
Regards Schabi

**Comment by andrewgsavage** [comment, marked as answer]
I think using apply will always be slower as it is applied row by row.

Can your function operate on an array?
```
v_2_g(df['voltage']...)
```
May just work

**Reply by SchabiDesigns** [reply]
@andrewgsavage thanks for your answer. 👍 
I checked that and I got an TypeError:
![image](https://github.com/hgrecco/pint/assets/97088571/617c67bd-b7a1-4ea7-8617-7f2970695f21)
